spec/index.bs

```diff
@@ -2883,57 +2964,79 @@ The {{GPUMapMode}} flags determine how a {{GPUBuffer}} is mapped when calling
+                        - |mode| contains exactly one of {{GPUMapMode/READ}} or {{GPUMapMode/WRITE}}.
+                        - If |mode| contains {{GPUMapMode/READ}} then |this|.{{GPUBuffer/usage}} must contain {{GPUBufferUsage/MAP_READ}}.
+                        - If |mode| contains {{GPUMapMode/WRITE}} then |this|.{{GPUBuffer/usage}} must contain {{GPUBufferUsage/MAP_WRITE}}.
+
+                        Issue: Do we validate that |mode| contains only valid flags?
+                    </div>
 
-                Then:
+                    Then:
 
-                1. Record a validation error on the current scope.
-                1. Return [=a promise rejected with=] an {{OperationError}} on the [=Device timeline=].
+                    1. Issue the following steps on |contentTimeline|.
 
-            1. Let |p| be a new {{Promise}}.
-            1. Set |this|.{{GPUBuffer/[[mapping]]}} to |p|.
-            1. Set |this|.{{GPUBuffer/[[state]]}} to [=buffer state/mapping pending=].
-            1. Set |this|.{{GPUBuffer/[[map_mode]]}} to |mode|.
-            1. Enqueue an operation on the default queue's [=Queue timeline=] that will execute the following:
+                        <ol data-timeline=content>
+                            1. [=Reject=] |p| with an {{OperationError}}.
+                        </ol>
 
-                <div class=queue-timeline>
-                    1. If |this|.{{GPUBuffer/[[state]]}} is [=buffer state/mapping pending=]:
+                    1. [$Generate a validation error$] and stop.
 
-                        1. Let |m| be a new {{ArrayBuffer}} of size |rangeSize|.
+                1. Once all <span data-timeline=queue>currently-enqueued operations (on any queue)</span>
+                    on |this| are completed, issue the |device timeline steps to map| on this timeline.
+            </ol>
 
-                            Note: This is just a memory allocation; this {{ArrayBuffer}} object is not exposed directly.
-                        1. Set the content of |m| to the content of |this|'s allocation starting at offset |offset| and for |rangeSize| bytes.
-                        1. Set |this|.{{GPUBuffer/[[mapping]]}} to |m|.
-                        1. Set |this|.{{GPUBuffer/[[state]]}} to [=buffer state/mapped=].
-                        1. Set |this|.{{GPUBuffer/[[mapping_range]]}} to <code>[|offset|, |offset| + |rangeSize|]</code>.
-                        1. Set |this|.{{GPUBuffer/[[mapped_ranges]]}} to `[]`.
+            The |device timeline steps to map| are:
 
-                    1. Resolve |p|.
-                </div>
-            1. Return |p|.
+            <ol data-timeline=device>
+                1. Issue: Set any state here as needed (unsure if any is needed actually).
+                1. Issue the |content timeline steps upon successful mapping| on |contentTimeline|.
+            </ol>
+
+            The |content timeline steps upon successful mapping| are:
```

Comment:
Given that the algorithm flows pretty directly from the previous timeline steps to the following content steps here, I'd prefer we find a way of indicating that transition that's less disjoint. As is this "Call the steps named X" pattern makes it feel harder to read. I think it's because I see the named steps and immediately equate that with a function of sorts, and begin wondering if it's called from multiple different places and thus I stop feeling like I can read the steps linearly.

I wonder if ending the device timeline steps here with something like "execute the next steps on the content timeline" and having the sections effectively share a border would read better? Or doing the same thing with the timeline borders but having the content section start with "Once the above steps have completed on the device timeline, continue with the following steps on the content timeline."